src/recommendation/src/main/scala/RankingTrainValidationSplit.scala

```diff
@@ -0,0 +1,343 @@
+      Future[Double] {
+        val model = est.fit(trainingDataset, paramMap).asInstanceOf[Model[_]]
+        calculateMetrics(model, validationDataset)
+      }(executionContext)
+    }
+
+    val metrics = metricFutures.map(SparkHelpers.getThreadUtils().awaitResult(_, Duration.Inf))
+
+    trainingDataset.unpersist()
+    validationDataset.unpersist()
+
+    val (bestMetric, bestIndex) =
+      if (eval.isLargerBetter) metrics.zipWithIndex.maxBy(_._1)
+      else metrics.zipWithIndex.minBy(_._1)
+
+    val bestModel = est.fit(dataset, epm(bestIndex)).asInstanceOf[Model[_]]
+    copyValues(new RankingTrainValidationSplitModel(uid)
+      .setBestModel(bestModel)
+      .setValidationMetrics(metrics)
+      .setParent(this))
+  }
+
+  override def copy(extra: ParamMap): RankingTrainValidationSplit = defaultCopy(extra)
+
+  private def filterByItemCount(dataset: Dataset[_]): DataFrame = {
+    dataset
+      .groupBy($(userCol))
+      .agg(col($(userCol)), count(col($(itemCol))))
+      .withColumnRenamed("count(" + $(itemCol) + ")", "nitems")
+      .where(col("nitems") >= $(minRatingsU))
+      .drop("nitems")
+      .cache()
+  }
+
+  private def filterByUserRatingCount(dataset: Dataset[_]): DataFrame = dataset
+    .groupBy($(itemCol))
+    .agg(col($(itemCol)), count(col($(userCol))))
+    .withColumnRenamed("count(" + $(userCol) + ")", "ncustomers")
+    .where(col("ncustomers") >= $(minRatingsI))
+    .join(dataset, $(itemCol))
+    .drop("ncustomers")
+    .cache()
+
+  def filterRatings(dataset: Dataset[_]): DataFrame = filterByUserRatingCount(dataset)
+    .join(filterByItemCount(dataset), $(userCol))
+
+  def splitDF(dataset: DataFrame): Array[DataFrame] = {
+    val shuffleFlag = true
+    val shuffleBC = dataset.sparkSession.sparkContext.broadcast(shuffleFlag)
+
+    if (dataset.columns.contains($(ratingCol))) {
+      val wrapColumn = udf((itemId: Double, rating: Double) => Array(itemId, rating))
+
+      val sliceudf = udf(
+        (r: mutable.WrappedArray[Array[Double]]) => r.slice(0, math.round(r.length * $(trainRatio)).toInt))
+
+      val shuffle = udf((r: mutable.WrappedArray[Array[Double]]) =>
+        if (shuffleBC.value) Random.shuffle(r.toSeq)
+        else r
+      )
+      val dropudf = udf((r: mutable.WrappedArray[Array[Double]]) => r.drop(math.round(r.length * $(trainRatio)).toInt))
+
+      val testds = dataset
+        .withColumn("itemIDRating", wrapColumn(col($(itemCol)), col($(ratingCol))))
+        .groupBy(col($(userCol)))
+        .agg(collect_list(col("itemIDRating")))
+        .withColumn("shuffle", shuffle(col("collect_list(itemIDRating)")))
+        .withColumn("train", sliceudf(col("shuffle")))
+        .withColumn("test", dropudf(col("shuffle")))
+        .drop(col("collect_list(itemIDRating)")).drop(col("shuffle"))
+        .cache()
+
+      val popLeft = udf((r: mutable.WrappedArray[Double]) => r(0))
+      val popRight = udf((r: mutable.WrappedArray[Double]) => r(1))
```

Comment:
can just use col("foo").getItem(0) or 1. This way spark will support query pushdown through it